In [913]:
from collections import defaultdict
from collections import Counter
import math

In [914]:
with open('day8.txt') as f:
    lines = [line for line in f]   

In [915]:
part_one = False

In [916]:
instructions = lines[0]
mapping = {"L":0,"R":1}
instructions = [mapping[i] for i in instructions if i in mapping.keys()]
#instructions

In [917]:
rules = defaultdict(tuple)
for rule in lines[2:]:
    key,value = rule.split(" = ")
    
    first_step = value.split("(")[1].split(",")
    value_one = first_step[0]
    value_two = first_step[1].split(")")[0][1:]
    rules[key]=(value_one,value_two)
#rules

In [918]:
finish_nodes = [i for i in rules.keys() if i[2]=="Z"]
finish_nodes

['BNZ', 'MGZ', 'BKZ', 'ZZZ', 'SSZ', 'TQZ']

In [919]:
class Walker:
    
    def __init__(self,initial_state):
        self.state = initial_state
        self.num_moves = 0
        self.status = "not finished"
                
    def update_state(self,move):
        self.state = rules[self.state][move]
        self.num_moves +=1
        
        if part_one:
            if self.state == "ZZZ":
                self.status = "finished"
                
        else:
            if self.state[2] == "Z":
                self.status = "finished"
            
        if part_one:
            if self.state != "ZZZ":
                self.status = "not finished"
        else:
            if self.state[2] != "Z":
                self.status = "not finished"

        
    
    def get_state(self):
        return self.state

In [920]:
if part_one:
    agent = Walker("AAA")
    status = "not finished"
    j=0
    while status != "finished":
        agent.update_state(instructions[j])
        if j< len(instructions)-1:
            j +=1
        else:
            j=0
        if agent.status == "finished":
            print("Finished in", agent.num_moves)
            break



In [921]:
starting_nodes = [i for i in rules.keys() if i[2]=="A"]
starting_nodes

['BFA', 'AAA', 'DFA', 'XFA', 'QJA', 'SBA']

In [922]:
finish_nodes = [i for i in rules.keys() if i[2]=="Z"]
finish_nodes

['BNZ', 'MGZ', 'BKZ', 'ZZZ', 'SSZ', 'TQZ']

In [923]:
#rules

In [924]:
len(instructions)

269

## Part 2


In [925]:
def get_num_moves(start,instruction_point):
    agent = Walker(start)
    status = "not finished"
    j=instruction_point
    
    while status != "finished":
        agent.update_state(instructions[j])
        if j< len(instructions)-1:
            j +=1
        else:
            j=0
        if agent.status == "finished":
            return(agent.num_moves,agent.state)
            break



In [948]:
# Distances to first finish node

starting_nodes = [i for i in rules.keys() if i[2]=="A"]


first_state = []
for start in starting_nodes:
    result = get_num_moves(start,0)
    first_state.append(result)
    



In [949]:
first_state = first_state
first_state

[(19637, 'MGZ'),
 (18023, 'ZZZ'),
 (21251, 'TQZ'),
 (16409, 'BKZ'),
 (11567, 'BNZ'),
 (14257, 'SSZ')]

## Notice that all first states are multiples of 269, the sequence repeats immediately for each starting position. So all we have to do is find LCM(n_1,n_2,...,n_6)

In [954]:
def lcm(a,b):
    return (a*b)/math.gcd(a,b)

In [955]:
numbers = [i[0] for i in first_state]

In [976]:
def lcm_list(input_list):
    answer = 1
    for i in input_list:
        answer = int(lcm(answer,i))
    return(answer)

In [977]:
lcm_list(numbers)

14449445933179